src/pybennu/pybennu/providers/power/solvers/siren.py

```diff
@@ -0,0 +1,493 @@
+"""
+SCEPTRE Provider for Siren
+
+## Overview
+
+Data is read from and written to predefined hardware (defined in the siren/hil_device directory)
+and published by this siren provider.
+
+"""
+
+import argparse
+import atexit
+import contextlib
+import importlib
+import inspect
+import inspect
+import json
+import logging
+import os
+import signal
+import sys
+import threading
+import time
+import math
+
+from typing import List, Literal, Tuple, Type, Union
+from collections import deque
+
+from pybennu.distributed.provider import Provider
+import pybennu.distributed.swig._Endpoint as E
+
+
+from time import sleep
+
+
+LEVELS = {'debug': logging.DEBUG,       # 10
+          'info': logging.INFO,         # 20
+          'warning': logging.WARNING,   # 30
+          'filter': 31,
+          'error': logging.ERROR,       # 40
+          'critical': logging.CRITICAL  # 50
+         }
+
+logging.addLevelName(LEVELS['filter'], 'FILTER')
+logging.basicConfig(format='%(asctime)s %(name)-12s  %(levelname)-8s  %(message)s', level=logging.DEBUG, datefmt='%Y-%m-%d %H:%M:%S')
+
+class CircularBuffer:
+    def __init__(self, max_size):
+        self.max = max_size
+        self.cur_size = 0
+        self.buffer = deque()
+        self.sum = 0
+
+    def add(self, value):
+        if len(self.buffer) == self.max:
+            self.sum -= self.buffer.popleft()
+
+        self.buffer.append(value)
+        self.sum += value
+        self.cur_size = len(self.buffer)
+            
```

Comment:
A single comment for a bunch of formatting things. basically, there should be one line between functions in a class, and two lines separating classes.